keras_nlp/src/utils/transformers/convert_llama3.py

```diff
@@ -0,0 +1,213 @@
+
+    Returns:
+        backbone: Initialized Keras model backbone.
+    """
+    transformers_config = load_config(preset, "config.json")
+
+    backbone = cls(
+        vocabulary_size=transformers_config["vocab_size"],
+        num_layers=transformers_config["num_hidden_layers"],
+        num_query_heads=transformers_config["num_attention_heads"],
+        hidden_dim=transformers_config["hidden_size"],
+        intermediate_dim=transformers_config["intermediate_size"],
+        num_key_value_heads=transformers_config["num_key_value_heads"],
+    )
+
+    if load_weights:
+        safetensor_config = load_config(preset, "model.safetensors.index.json")
+        safetensor_files = {
+            fname: get_file(preset, fname)
+            for fname in set(safetensor_config["weight_map"].values())
+        }
+        port_weight = partial(
+            set_keras_weight,
+            safetensor_files=safetensor_files,
+            safetensor_config=safetensor_config,
+        )
+
+        # Embedding layers
+        port_weight(
+            keras_variable=backbone.get_layer("token_embedding").variables[0],
+            hf_weight_key="model.embed_tokens.weight",
+        )
+        port_weight(
+            keras_variable=backbone.get_layer("token_embedding").variables[1],
+            hf_weight_key="lm_head.weight",
+            # rearrange_pattern="b a -> a b",
+            hook_fn=lambda hf_tensor, _: np.transpose(hf_tensor, axes=(1, 0)),
+        )
+
+        # Attention blocks
+        for i in range(backbone.num_layers):
+            decoder_layer = backbone.get_layer(f"transformer_layer_{i}")
+            # Norm layers
+            port_weight(
+                keras_variable=decoder_layer._self_attention_layernorm.variables[
+                    0
+                ],
+                hf_weight_key=f"model.layers.{i}.input_layernorm.weight",
+            )
+            port_weight(
+                keras_variable=decoder_layer._feedforward_layernorm.variables[
+                    0
+                ],
+                hf_weight_key=f"model.layers.{i}.post_attention_layernorm.weight",
+            )
+
+            # Attention layers
+            port_weight(
+                keras_variable=decoder_layer._self_attention_layer._query_dense.variables[
+                    0
+                ],
+                hf_weight_key=f"model.layers.{i}.self_attn.q_proj.weight",
+                # rearrange_patterns="(b c) a -> a b c,
+                # rearrange_dims={"b": backbone.num_query_heads},
+                hook_fn=lambda hf_tensor, keras_shape: np.transpose(
+                    np.reshape(
+                        hf_tensor,
+                        (keras_shape[1], keras_shape[2], keras_shape[0]),
+                    ),
+                    axes=(2, 0, 1),
+                ),
+            )
+            port_weight(
+                keras_variable=decoder_layer._self_attention_layer._key_dense.variables[
```

Comment:
General comment for all of these, can't you use `kernel` instead of `variables[0]`? And `bias`, etc. The variable names will be a lot more readable.

Reply:
The only reason I was shying away from using `kernel` and `bias` is because, it is not consistent with all the layers.
An embedding layer has `embeddings`, a normalization layer has `scale` and a dense layer has `kernel`. If you want me to make the changes, I would need to run the scripts and be sure about the variable names.

What would you like me to do?

Reply:
All good to leave as is then! Maybe something we can do as a follow up. Should be pretty easy to do safely now that we have testing in place.